src/deepsparse/pipeline.py

```diff
@@ -176,23 +174,120 @@ def __init__(
+            postprocess_kwargs = {}
 
-        return (
-            executor.submit(self._run, *args, **kwargs)  # Non-Blocking call
-            if executor and not self.use_dynamic_batch()
-            else self._run(*args, **kwargs)  # Blocking call
+        # split inputs into batches of size `self._batch_size`
+        batches = self.split_engine_inputs(engine_inputs, self._batch_size)
+
+        # submit to engine
+        futures = [
+            self.executor.submit(self.engine_forward, batch) for batch in batches
+        ]
+        wait(futures)
+
+        # join together the batches of size `self._batch_size`
+        engine_outputs = self.join_engine_outputs(
+            [future.result() for future in futures]
         )
 
+        pipeline_outputs = self.process_engine_outputs(
+            engine_outputs, **postprocess_kwargs
+        )
+        if not isinstance(pipeline_outputs, self.output_schema):
+            raise ValueError(
+                f"Outputs of {self.__class__} must be instances of "
+                f"{self.output_schema} found output of type {type(pipeline_outputs)}"
+            )
+
+        return pipeline_outputs
+
+    @staticmethod
+    def split_engine_inputs(
+        items: List[numpy.ndarray], batch_size: int
+    ) -> List[List[numpy.ndarray]]:
+        """
+        Splits each item into numpy arrays with the first dimension == `batch_size`.
+
+        For example, if `items` has three numpy arrays with the following
+        shapes: `[(4, 32, 32), (4, 64, 64), (4, 128, 128)]`
+
+        Then with `batch_size==4` the output would be:
+        ```
+        [[(4, 32, 32), (4, 64, 64), (4, 128, 128)]]
+        ```
+
+        Then with `batch_size==2` the output would be:
+        ```
+        [
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+            [(2, 32, 32), (2, 64, 64), (2, 128, 128)],
+        ]
+        ```
+
+        Then with `batch_size==1` the output would be:
+        ```
+        [
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+            [(1, 32, 32), (1, 64, 64), (1, 128, 128)],
+        ]
+        ```
+        """
+        # if not all items here are numpy arrays, there's an internal
+        # but in the processing code
+        assert all(isinstance(item, numpy.ndarray) for item in items)
+
+        # if not all items have the same batch size, there's an
+        # internal bug in the processing code
+        total_batch_size = items[0].shape[0]
+        assert all(item.shape[0] == total_batch_size for item in items)
+
+        if total_batch_size % batch_size != 0:
+            raise RuntimeError(
+                f"batch size of {total_batch_size} passed into pipeline "
+                f"is not divisible by model batch size of {batch_size}"
+            )
+
+        batches = []
+        for i_batch in range(total_batch_size // batch_size):
```

Comment:
nit: range also has a step argument

Reply:
Oooh nice one, are you thinking `range(0, total_batch_size, batch_size)`?